### ISSUE TITLE
Expected Results for PyTorch run_lm1b_base.sh

### ISSUE THREAD

**Issue by rdspring1** [kimiyoung/transformer-xl#69]
I wanted to know the expected test perplexity for the lm1b base model. It would be especially great if you could upload the training log file if possible. I wanted to include the results in my ICML paper.

**Comment by kimiyoung** [comment]
We don't have numbers for that particular setting. In our paper, we used a base model with N_LAYER=20, D_MODEL=1024, D_INNER=6144 for lm1b and obtained a test perplexity of 23.5 (See Table 4 of https://arxiv.org/pdf/1901.02860.pdf).

**Comment by rdspring1** [comment]
For the pytorch script, the lm1b base model uses N_LAYER=18, D_MODEL=1024, D_INNER=4096 (about 0.36 billion parameters). After 500K training step, my test perplexity was 30.25. My work uses a memory compressed version of the Adam optimizer. Would you consider this test score unreasonable?

**Comment by kimiyoung** [comment]
I would expect a (much) better perplexity given the model size.

**Comment by rdspring1** [comment]
Then, do you have the log files for the original LM1B base model in the paper?
I'll try using those settings.

In this issue [https://github.com/kimiyoung/transformer-xl/issues/1](https://github.com/kimiyoung/transformer-xl/issues/1), you said that you ran the base models in PyTorch.

**Comment by kimiyoung** [comment]
No I don't have access to the log files at this point. The base model results in Table 4 of our paper were also produced using TF. Try `N_LAYER=20, D_MODEL=1024, D_INNER=6144` and I believe the pytorch version will also likely get close to 23.5. As for `N_LAYER=18, D_MODEL=1024, D_INNER=4096`, have you run the model as is to see the performance?

**Comment by rdspring1** [comment]
When I use the run_lm1b_base script I see this? However, this behavior doesn't seem correct - oscillating around 65 perplexity.
```
#params = 357812065
#non emb params = 245532672
----------------------------------------------------------------------------------------------------
| Eval  12 at step    48000 | time: 5132.39s | valid loss  4.16 | valid ppl    64.240
----------------------------------------------------------------------------------------------------
| epoch   1 step    48200 |  48200 batches | lr 0.000244 | ms/batch 1410.03 | loss  4.13 | ppl    61.889
| epoch   1 step    48400 |  48400 batches | lr 0.000244 | ms/batch 1294.38 | loss  4.14 | ppl    62.738
| epoch   1 step    48600 |  48600 batches | lr 0.000244 | ms/batch 1266.15 | loss  4.17 | ppl    64.460
| epoch   1 step    48800 |  48800 batches | lr 0.000244 | ms/batch 1268.40 | loss  4.18 | ppl    65.431
| epoch   1 step    49000 |  49000 batches | lr 0.000244 | ms/batch 1273.59 | loss  4.19 | ppl    66.287
| epoch   1 step    49200 |  49200 batches | lr 0.000244 | ms/batch 1266.28 | loss  4.18 | ppl    65.394
| epoch   1 step    49400 |  49400 batches | lr 0.000244 | ms/batch 1264.99 | loss  4.18 | ppl    65.567
| epoch   1 step    49600 |  49600 batches | lr 0.000244 | ms/batch 1302.28 | loss  4.20 | ppl    66.799
| epoch   1 step    49800 |  49800 batches | lr 0.000244 | ms/batch 1266.97 | loss  4.20 | ppl    66.702
| epoch   1 step    50000 |  50000 batches | lr 0.000244 | ms/batch 1267.11 | loss  4.16 | ppl    64.154
| epoch   1 step    50200 |  50200 batches | lr 0.000244 | ms/batch 1266.39 | loss  4.14 | ppl    62.508
| epoch   1 step    50400 |  50400 batches | lr 0.000244 | ms/batch 1266.90 | loss  4.16 | ppl    63.760
| epoch   1 step    50600 |  50600 batches | lr 0.000244 | ms/batch 1265.40 | loss  4.15 | ppl    63.486
| epoch   1 step    50800 |  50800 batches | lr 0.000244 | ms/batch 1289.09 | loss  4.14 | ppl    62.990
| epoch   1 step    51000 |  51000 batches | lr 0.000244 | ms/batch 1270.12 | loss  4.14 | ppl    62.907
| epoch   1 step    51200 |  51200 batches | lr 0.000244 | ms/batch 1266.82 | loss  4.16 | ppl    64.348
| epoch   1 step    51400 |  51400 batches | lr 0.000244 | ms/batch 1266.81 | loss  4.15 | ppl    63.356
| epoch   1 step    51600 |  51600 batches | lr 0.000243 | ms/batch 1266.01 | loss  4.15 | ppl    63.264
| epoch   1 step    51800 |  51800 batches | lr 0.000243 | ms/batch 1297.72 | loss  4.17 | ppl    64.703
| epoch   1 step    52000 |  52000 batches | lr 0.000243 | ms/batch 1270.37 | loss  4.16 | ppl    64.097
----------------------------------------------------------------------------------------------------
| Eval  13 at step    52000 | time: 5119.55s | valid loss  4.13 | valid ppl    61.921
----------------------------------------------------------------------------------------------------
| epoch   1 step    52200 |  52200 batches | lr 0.000243 | ms/batch 1402.27 | loss  4.17 | ppl    64.870
| epoch   1 step    52400 |  52400 batches | lr 0.000243 | ms/batch 1270.55 | loss  4.17 | ppl    64.691
| epoch   1 step    52600 |  52600 batches | lr 0.000243 | ms/batch 1266.28 | loss  4.16 | ppl    64.062
| epoch   1 step    52800 |  52800 batches | lr 0.000243 | ms/batch 1266.92 | loss  4.16 | ppl    64.056
| epoch   1 step    53000 |  53000 batches | lr 0.000243 | ms/batch 1291.17 | loss  4.26 | ppl    70.585
| epoch   1 step    53200 |  53200 batches | lr 0.000243 | ms/batch 1271.03 | loss  4.20 | ppl    66.751
| epoch   1 step    53400 |  53400 batches | lr 0.000243 | ms/batch 1271.76 | loss  4.16 | ppl    64.147
| epoch   1 step    53600 |  53600 batches | lr 0.000243 | ms/batch 1284.31 | loss  4.18 | ppl    65.402
| epoch   1 step    53800 |  53800 batches | lr 0.000243 | ms/batch 1271.13 | loss  4.19 | ppl    65.710
| epoch   1 step    54000 |  54000 batches | lr 0.000243 | ms/batch 1294.83 | loss  4.17 | ppl    65.039
| epoch   1 step    54200 |  54200 batches | lr 0.000243 | ms/batch 1271.26 | loss  4.17 | ppl    64.620
| epoch   1 step    54400 |  54400 batches | lr 0.000243 | ms/batch 1269.98 | loss  4.18 | ppl    65.041
| epoch   1 step    54600 |  54600 batches | lr 0.000243 | ms/batch 1271.97 | loss  4.17 | ppl    64.832
| epoch   1 step    54800 |  54800 batches | lr 0.000243 | ms/batch 1271.84 | loss  4.16 | ppl    64.159
| epoch   1 step    55000 |  55000 batches | lr 0.000243 | ms/batch 1270.68 | loss  4.16 | ppl    64.344
| epoch   1 step    55200 |  55200 batches | lr 0.000243 | ms/batch 1294.90 | loss  4.17 | ppl    64.787
| epoch   1 step    55400 |  55400 batches | lr 0.000243 | ms/batch 1280.10 | loss  4.20 | ppl    66.789
| epoch   1 step    55600 |  55600 batches | lr 0.000242 | ms/batch 1269.64 | loss  4.23 | ppl    68.416
| epoch   1 step    55800 |  55800 batches | lr 0.000242 | ms/batch 1272.26 | loss  4.20 | ppl    66.791
| epoch   1 step    56000 |  56000 batches | lr 0.000242 | ms/batch 1270.36 | loss  4.21 | ppl    67.023
----------------------------------------------------------------------------------------------------
| Eval  14 at step    56000 | time: 5126.69s | valid loss  4.18 | valid ppl    65.364
----------------------------------------------------------------------------------------------------
| epoch   1 step    56200 |  56200 batches | lr 0.000242 | ms/batch 1400.62 | loss  4.20 | ppl    66.641
| epoch   1 step    56400 |  56400 batches | lr 0.000242 | ms/batch 1294.03 | loss  4.18 | ppl    65.534
| epoch   1 step    56600 |  56600 batches | lr 0.000242 | ms/batch 1278.80 | loss  4.18 | ppl    65.627
| epoch   1 step    56800 |  56800 batches | lr 0.000242 | ms/batch 1271.03 | loss  4.22 | ppl    67.794
| epoch   1 step    57000 |  57000 batches | lr 0.000242 | ms/batch 1271.45 | loss  4.17 | ppl    64.446
| epoch   1 step    57200 |  57200 batches | lr 0.000242 | ms/batch 1280.95 | loss  4.20 | ppl    66.547
| epoch   1 step    57400 |  57400 batches | lr 0.000242 | ms/batch 1296.23 | loss  4.16 | ppl    64.168
| epoch   1 step    57600 |  57600 batches | lr 0.000242 | ms/batch 1278.13 | loss  4.17 | ppl    64.582
| epoch   1 step    57800 |  57800 batches | lr 0.000242 | ms/batch 1269.62 | loss  4.16 | ppl    64.314
| epoch   1 step    58000 |  58000 batches | lr 0.000242 | ms/batch 1284.13 | loss  4.14 | ppl    62.884
| epoch   1 step    58200 |  58200 batches | lr 0.000242 | ms/batch 1284.32 | loss  4.17 | ppl    64.793
| epoch   1 step    58400 |  58400 batches | lr 0.000242 | ms/batch 1272.11 | loss  4.17 | ppl    64.411
| epoch   1 step    58600 |  58600 batches | lr 0.000242 | ms/batch 1293.54 | loss  4.18 | ppl    65.501
| epoch   1 step    58800 |  58800 batches | lr 0.000242 | ms/batch 1277.37 | loss  4.20 | ppl    66.846
| epoch   1 step    59000 |  59000 batches | lr 0.000242 | ms/batch 1275.10 | loss  4.20 | ppl    66.643
| epoch   1 step    59200 |  59200 batches | lr 0.000241 | ms/batch 1270.24 | loss  4.19 | ppl    66.168
| epoch   1 step    59400 |  59400 batches | lr 0.000241 | ms/batch 1281.45 | loss  4.20 | ppl    66.690
| epoch   1 step    59600 |  59600 batches | lr 0.000241 | ms/batch 1269.60 | loss  4.18 | ppl    65.125
| epoch   1 step    59800 |  59800 batches | lr 0.000241 | ms/batch 1290.70 | loss  4.16 | ppl    64.041
| epoch   1 step    60000 |  60000 batches | lr 0.000241 | ms/batch 1269.11 | loss  4.20 | ppl    66.848
----------------------------------------------------------------------------------------------------
| Eval  15 at step    60000 | time: 5141.84s | valid loss  4.23 | valid ppl    68.582
----------------------------------------------------------------------------------------------------
```

**Comment by kimiyoung** [comment]
It has been a known issue that using `DIV_VAL > 1` (i.e., using adaptive embeddings) might cause instability of optimization (See #24). Try decreasing `init_std` or increasing the number of warmup steps.

**Comment by cmathx** [comment]
@kimiyoung I ran run_lm1b_base script in 16GB memory nvidia gpu, but it occurs OOM. When I decrease BSZ from 256 to 64, it ran well. But I think convergence of this configuration may cost 4 times of the time-cost. Whether any variable can be stored in cpu not gpu?

**Comment by cmathx** [comment]
> For the pytorch script, the lm1b base model uses N_LAYER=18, D_MODEL=1024, D_INNER=4096 (about 0.36 billion parameters). After 500K training step, my test perplexity was 30.25. My work uses a memory compressed version of the Adam optimizer. Would you consider this test score unreasonable?
The test set perplexity which is 30.25 is got by the default pytorch script(run_lm1b_base script) only modified the gradient optimizer(a memory compressed version of the Adam optimizer)? Did you modify other hyper parameters?
